File: c/bindle.c
#include "bindle.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/file.h>
#include <zstd.h>

#define BNDL_MAGIC "BINDL001"
#define BNDL_ALIGN 8
#define ALIGN_UP(n, m) (((n) + (m) - 1) & ~((m) - 1))

/* --- Private Disk Structures --- */
#pragma pack(push, 1)
typedef struct {
  uint64_t offset;
  uint64_t compressed_size;
  uint64_t uncompressed_size;
  uint32_t crc32;
  uint16_t name_len;
  uint8_t compression_type;
  uint8_t _reserved;
} BindleEntryRaw;

typedef struct {
  uint64_t index_offset;
  uint64_t entry_count;
} BindleFooterRaw;
#pragma pack(pop)

/* --- Private In-Memory Structures --- */
typedef struct {
  BindleEntryRaw meta;
  char *name;
} BindleEntry;

struct Bindle {
  char *path;
  FILE *fp;
  BindleEntry *entries;
  uint64_t count;
  uint64_t data_end;
};

/* --- API Implementation --- */

Bindle *bindle_open(const char *path) {
  FILE *fp = fopen(path, "r+b");
  if (!fp) {
    fp = fopen(path, "w+b");
  }
  if (!fp)
    return NULL;

  flock(fileno(fp), LOCK_SH);

  Bindle *b = calloc(1, sizeof(Bindle));
  b->path = strdup(path);
  b->fp = fp;

  fseek(fp, 0, SEEK_END);
  long file_size = ftell(fp);

  if (file_size == 0) {
    fwrite(BNDL_MAGIC, 8, 1, fp);
    b->data_end = 8;
    return b;
  }

  // Header check
  char magic[8];
  fseek(fp, 0, SEEK_SET);
  if (fread(magic, 8, 1, fp) != 1 || memcmp(magic, BNDL_MAGIC, 8) != 0) {
    bindle_close(b);
    return NULL;
  }

  // Footer parse
  BindleFooterRaw footer;
  fseek(fp, file_size - sizeof(BindleFooterRaw), SEEK_SET);
  if (fread(&footer, sizeof(BindleFooterRaw), 1, fp) != 1) {
    bindle_close(b);
    return NULL;
  }

  b->count = footer.entry_count;
  b->data_end = footer.index_offset;
  b->entries = malloc(sizeof(BindleEntry) * b->count);

  // Index parse
  fseek(fp, footer.index_offset, SEEK_SET);
  for (uint64_t i = 0; i < b->count; i++) {
    fread(&b->entries[i].meta, sizeof(BindleEntryRaw), 1, fp);
    b->entries[i].name = malloc(b->entries[i].meta.name_len + 1);
    fread(b->entries[i].name, b->entries[i].meta.name_len, 1, fp);
    b->entries[i].name[b->entries[i].meta.name_len] = '\0';

    size_t consumed = sizeof(BindleEntryRaw) + b->entries[i].meta.name_len;
    fseek(fp, ALIGN_UP(consumed, BNDL_ALIGN) - consumed, SEEK_CUR);
  }
  return b;
}

bool bindle_add(Bindle *b, const char *name, const uint8_t *data, size_t len,
                BindleCompress compress) {
  if (!b || !name)
    return false;

  size_t c_size = len;
  void *write_ptr = (void *)data;
  void *comp_buf = NULL;

  if (compress == BindleCompressZstd) {
    size_t bound = ZSTD_compressBound(len);
    comp_buf = malloc(bound);
    c_size = ZSTD_compress(comp_buf, bound, data, len, 3);
    if (ZSTD_isError(c_size)) {
      free(comp_buf);
      return false;
    }
    write_ptr = comp_buf;
  }

  // 1. Write data at current data_end
  fseek(b->fp, b->data_end, SEEK_SET);
  uint64_t offset = ftell(b->fp);
  fwrite(write_ptr, 1, c_size, b->fp);

  // 2. Align data_end
  size_t pad = ALIGN_UP(c_size, BNDL_ALIGN) - c_size;
  if (pad > 0) {
    uint8_t zero[8] = {0};
    fwrite(zero, 1, pad, b->fp);
  }
  b->data_end = ftell(b->fp);

  // 3. Shadowing: Check if name already exists
  for (uint64_t i = 0; i < b->count; i++) {
    if (strcmp(b->entries[i].name, name) == 0) {
      b->entries[i].meta.offset = offset;
      b->entries[i].meta.compressed_size = c_size;
      b->entries[i].meta.uncompressed_size = len;
      b->entries[i].meta.compression_type = compress ? 1 : 0;
      if (comp_buf)
        free(comp_buf);
      return true;
    }
  }

  // 4. New Entry
  b->entries = realloc(b->entries, sizeof(BindleEntry) * (b->count + 1));
  BindleEntry *e = &b->entries[b->count++];
  e->name = strdup(name);
  e->meta = (BindleEntryRaw){offset,   c_size, len, 0, (uint16_t)strlen(name),
                             compress, 0};

  if (comp_buf)
    free(comp_buf);
  return true;
}

uint8_t *bindle_read(Bindle *b, const char *name, size_t *out_len) {
  for (uint64_t i = 0; i < b->count; i++) {
    if (strcmp(b->entries[i].name, name) == 0) {
      BindleEntryRaw *m = &b->entries[i].meta;
      uint8_t *c_buf = malloc(m->compressed_size);
      fseek(b->fp, m->offset, SEEK_SET);
      fread(c_buf, 1, m->compressed_size, b->fp);

      if (m->compression_type == BindleCompressZstd) {
        uint8_t *u_buf = malloc(m->uncompressed_size);
        size_t actual = ZSTD_decompress(u_buf, m->uncompressed_size, c_buf,
                                        m->compressed_size);
        free(c_buf);
        *out_len = actual;
        return u_buf;
      }
      *out_len = m->compressed_size;
      return c_buf;
    }
  }
  return NULL;
}

const uint8_t *bindle_read_uncompressed_direct(Bindle *b, const char *name,
                                               size_t *out_len) {
  for (uint64_t i = 0; i < b->count; i++) {
    if (strcmp(b->entries[i].name, name) == 0) {
      BindleEntryRaw *m = &b->entries[i].meta;
      if (m->compression_type != BindleCompressNone)
        return NULL;

      uint8_t *buf = malloc(m->uncompressed_size);
      fseek(b->fp, m->offset, SEEK_SET);
      fread(buf, 1, m->uncompressed_size, b->fp);
      *out_len = m->uncompressed_size;
      return buf;
    }
  }
  return NULL;
}

bool bindle_save(Bindle *b) {
  flock(fileno(b->fp), LOCK_EX);
  fseek(b->fp, b->data_end, SEEK_SET);
  uint64_t index_start = b->data_end;

  for (uint64_t i = 0; i < b->count; i++) {
    fwrite(&b->entries[i].meta, sizeof(BindleEntryRaw), 1, b->fp);
    fwrite(b->entries[i].name, 1, b->entries[i].meta.name_len, b->fp);
    size_t consumed = sizeof(BindleEntryRaw) + b->entries[i].meta.name_len;
    size_t pad = ALIGN_UP(consumed, BNDL_ALIGN) - consumed;
    if (pad > 0) {
      uint8_t zero[8] = {0};
      fwrite(zero, 1, pad, b->fp);
    }
  }

  BindleFooterRaw footer = {index_start, b->count};
  fwrite(&footer, sizeof(BindleFooterRaw), 1, b->fp);
  fflush(b->fp);
  flock(fileno(b->fp), LOCK_SH);
  return true;
}

size_t bindle_length(const Bindle *b) { return b ? b->count : 0; }

const char *bindle_entry_name(const Bindle *b, size_t index, size_t *namelen) {
  if (!b || index >= b->count)
    return NULL;
  *namelen = b->entries[index].meta.name_len;
  return b->entries[index].name;
}

void bindle_free_buffer(uint8_t *ptr) { free(ptr); }

void bindle_close(Bindle *b) {
  if (!b)
    return;
  flock(fileno(b->fp), LOCK_UN);
  for (uint64_t i = 0; i < b->count; i++)
    free(b->entries[i].name);
  free(b->entries);
  fclose(b->fp);
  free(b->path);
  free(b);
}

bool bindle_vacuum(Bindle *b) {
  if (!b)
    return false;

  char tmp_path[1024];
  snprintf(tmp_path, sizeof(tmp_path), "%s.tmp", b->path);
  FILE *out = fopen(tmp_path, "wb");
  if (!out)
    return false;

  // 1. Write Header
  fwrite(BNDL_MAGIC, 8, 1, out);
  uint64_t current_offset = 8;

  // 2. Copy Live Data to Temp File
  for (uint64_t i = 0; i < b->count; i++) {
    uint64_t size = b->entries[i].meta.compressed_size;
    uint8_t *buf = malloc(size);

    fseek(b->fp, b->entries[i].meta.offset, SEEK_SET);
    fread(buf, 1, size, b->fp);

    fseek(out, current_offset, SEEK_SET);
    fwrite(buf, 1, size, out);

    // Update the in-memory metadata with the new offset
    b->entries[i].meta.offset = current_offset;

    size_t pad = ALIGN_UP(size, BNDL_ALIGN) - size;
    if (pad > 0) {
      uint8_t zero[8] = {0};
      fwrite(zero, 1, pad, out);
    }
    current_offset += size + pad;
    free(buf);
  }

  // 3. Write Index and Footer to the Temp File (Matching Rust .save() logic)
  uint64_t index_start = current_offset;
  for (uint64_t i = 0; i < b->count; i++) {
    fwrite(&b->entries[i].meta, sizeof(BindleEntryRaw), 1, out);
    fwrite(b->entries[i].name, 1, b->entries[i].meta.name_len, out);

    size_t consumed = sizeof(BindleEntryRaw) + b->entries[i].meta.name_len;
    size_t pad = ALIGN_UP(consumed, BNDL_ALIGN) - consumed;
    if (pad > 0) {
      uint8_t zero[8] = {0};
      fwrite(zero, 1, pad, out);
    }
  }

  BindleFooterRaw footer = {index_start, b->count};
  fwrite(&footer, sizeof(BindleFooterRaw), 1, out);

  // 4. CRITICAL: Close and Unlock handles before Rename
  fflush(out);
  fclose(out); // Close the temp file handle

  flock(fileno(b->fp), LOCK_UN); // Explicitly unlock the original file
  fclose(b->fp);                 // Close the original file handle
  b->fp = NULL;

  // 5. Atomic Rename
  if (rename(tmp_path, b->path) != 0) {
    // If rename fails, we are in a bad state; attempt to re-open original
    b->fp = fopen(b->path, "r+b");
    return false;
  }

  // 6. Re-open the new primary file
  b->fp = fopen(b->path, "r+b");
  if (!b->fp)
    return false;

  flock(fileno(b->fp), LOCK_SH);
  b->data_end = index_start;

  return true;
}

File: src/bindle.rs
use crc32fast::Hasher;
use fs2::FileExt;
use memmap2::Mmap;
use std::borrow::Cow;
use std::collections::BTreeMap;
use std::fs::{File, OpenOptions};
use std::io::{self, Read, Seek, SeekFrom, Write};
use std::path::{Path, PathBuf};
use zerocopy::{FromBytes, IntoBytes};

use crate::compress::Compress;
use crate::entry::{Entry, Footer};
use crate::reader::{Either, Reader};
use crate::writer::Writer;
use crate::{
    AUTO_COMPRESS_THRESHOLD, BNDL_ALIGN, BNDL_MAGIC, ENTRY_SIZE, FOOTER_MAGIC, FOOTER_SIZE,
    HEADER_SIZE, pad, write_padding,
};

/// A binary archive for collecting files.
///
/// Uses memory-mapped I/O for fast reads, supports optional zstd compression, and handles updates via shadowing.
/// Files can be added incrementally without rewriting the entire archive.
///
/// # Example
///
/// ```no_run
/// use bindle_file::{Bindle, Compress};
///
/// let mut archive = Bindle::open("data.bndl")?;
/// archive.add("file.txt", b"data", Compress::None)?;
/// archive.save()?;
/// # Ok::<(), std::io::Error>(())
/// ```
pub struct Bindle {
    pub(crate) path: PathBuf,
    pub(crate) file: File,
    pub(crate) mmap: Option<Mmap>,
    pub(crate) index: BTreeMap<String, Entry>,
    pub(crate) data_end: u64,
}

impl Bindle {
    /// Creates a new archive, overwriting any existing file at the path.
    pub fn create<P: AsRef<Path>>(path: P) -> io::Result<Self> {
        let path_buf = path.as_ref().to_path_buf();
        let opts = OpenOptions::new()
            .truncate(true)
            .read(true)
            .write(true)
            .create(true)
            .to_owned();
        Self::new(path_buf, opts)
    }

    /// Opens an existing archive or creates a new one if it doesn't exist.
    pub fn open<P: AsRef<Path>>(path: P) -> io::Result<Self> {
        let path_buf = path.as_ref().to_path_buf();
        let opts = OpenOptions::new()
            .read(true)
            .write(true)
            .create(true)
            .to_owned();
        Self::new(path_buf, opts)
    }

    /// Opens an existing archive. Returns an error if the file doesn't exist.
    pub fn load<P: AsRef<Path>>(path: P) -> io::Result<Self> {
        let path_buf = path.as_ref().to_path_buf();
        let opts = OpenOptions::new().read(true).write(true).to_owned();
        Self::new(path_buf, opts)
    }

    /// Create a new `Bindle` from a path and file, the path must match the file
    pub fn new(path: PathBuf, opts: OpenOptions) -> io::Result<Self> {
        let mut file = opts.open(&path)?;
        file.lock_shared()?;
        let len = file.metadata()?.len();

        // Handle completely new/empty files
        if len == 0 {
            file.write_all(BNDL_MAGIC)?;
            return Ok(Self {
                path,
                file,
                mmap: None,
                index: BTreeMap::new(),
                data_end: HEADER_SIZE as u64,
            });
        }

        // Safety check: File must be at least HEADER + FOOTER size (24 bytes)
        // This prevents "attempt to subtract with overflow" when calculating footer_pos
        if len < (HEADER_SIZE + FOOTER_SIZE) as u64 {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "File too small to be a valid bindle",
            ));
        }

        let mut header = [0u8; 8];
        file.read_exact(&mut header)?;
        if &header != BNDL_MAGIC {
            return Err(io::Error::new(io::ErrorKind::InvalidData, "Invalid header"));
        }

        let m = unsafe { Mmap::map(&file)? };

        // Calculate footer position. Subtraction is now safe due to the check above.
        let footer_pos = m.len() - FOOTER_SIZE;
        let footer = Footer::read_from_bytes(&m[footer_pos..])
            .map_err(|_| io::Error::new(io::ErrorKind::InvalidData, "Failed to read footer"))?;

        if footer.magic() != FOOTER_MAGIC {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "Invalid footer, the file may be corrupt",
            ));
        }

        let data_end = footer.index_offset();
        let count = footer.entry_count();
        let mut index = BTreeMap::new();

        let mut cursor = data_end as usize;
        for _ in 0..count {
            // Ensure there is enough data left for an Entry header
            if cursor + ENTRY_SIZE > footer_pos {
                break;
            }

            let entry = match Entry::read_from_bytes(&m[cursor..cursor + ENTRY_SIZE]) {
                Ok(e) => e,
                Err(_) => break, // Corrupted entry, stop reading
            };
            let n_start = cursor + ENTRY_SIZE;

            // Validate that the filename exists within the mapped bounds
            if n_start + entry.name_len() > footer_pos {
                break;
            }

            let name =
                String::from_utf8_lossy(&m[n_start..n_start + entry.name_len()]).into_owned();
            index.insert(name, entry);

            let total = ENTRY_SIZE + entry.name_len();
            cursor += (total + (BNDL_ALIGN - 1)) & !(BNDL_ALIGN - 1);
        }

        Ok(Self {
            path,
            file,
            mmap: Some(m),
            index,
            data_end,
        })
    }

    fn should_auto_compress(&self, compress: Compress, len: usize) -> bool {
        compress == Compress::Zstd || (compress == Compress::Auto && len > AUTO_COMPRESS_THRESHOLD)
    }

    /// Adds data to the archive with the given name.
    ///
    /// If an entry with the same name exists, it will be shadowed. Call [`save()`](Bindle::save) to commit changes.
    pub fn add(&mut self, name: &str, data: &[u8], compress: Compress) -> io::Result<()> {
        let mut stream = self.writer(name, compress)?;
        stream.write_all(data)?;
        stream.close()?;
        Ok(())
    }

    /// Adds a file from the filesystem to the archive.
    ///
    /// Reads the file at `path` and stores it with the given `name`. Call [`save()`](Bindle::save) to commit changes.
    pub fn add_file(
        &mut self,
        name: &str,
        path: impl AsRef<Path>,
        compress: Compress,
    ) -> io::Result<()> {
        let mut stream = self.writer(name, compress)?;
        let mut src = std::fs::File::open(path)?;
        std::io::copy(&mut src, &mut stream)?;
        Ok(())
    }

    /// Commits all pending changes by writing the index and footer to disk.
    ///
    /// Must be called after add/remove operations to make changes persistent.
    pub fn save(&mut self) -> io::Result<()> {
        self.file.lock_exclusive()?;
        self.file.seek(SeekFrom::Start(self.data_end))?;
        let index_start = self.data_end;

        for (name, entry) in &self.index {
            self.file.write_all(entry.as_bytes())?;
            self.file.write_all(name.as_bytes())?;
            let pad = pad::<BNDL_ALIGN, usize>(ENTRY_SIZE + name.len());
            if pad > 0 {
                write_padding(&mut self.file, pad)?;
            }
        }

        let footer = Footer::new(index_start, self.index.len() as u32, FOOTER_MAGIC);
        self.file.write_all(footer.as_bytes())?;

        // Truncate file to current position to remove any old data
        let current_pos = self.file.stream_position()?;
        self.file.set_len(current_pos)?;
        self.file.flush()?;

        self.mmap = Some(unsafe { Mmap::map(&self.file)? });
        self.file.lock_shared()?;
        Ok(())
    }

    /// Reclaims space by removing shadowed data.
    ///
    /// Rebuilds the archive with only live entries, removing old versions of updated files.
    pub fn vacuum(&mut self) -> io::Result<()> {
        let temp_path = self.path.with_extension("tmp");

        // Create temp file and keep handle to reuse after rename
        let mut temp_file = OpenOptions::new()
            .write(true)
            .read(true)
            .create(true)
            .truncate(true)
            .open(&temp_path)?;

        temp_file.lock_exclusive()?;
        temp_file.write_all(BNDL_MAGIC)?;
        let mut current_offset = HEADER_SIZE as u64;

        // Copy only live entries from original to temp
        for entry in self.index.values_mut() {
            self.file.seek(SeekFrom::Start(entry.offset()))?;
            temp_file.seek(SeekFrom::Start(current_offset))?;

            // Stream data without allocating full buffer
            let mut limited = (&mut self.file).take(entry.compressed_size());
            io::copy(&mut limited, &mut temp_file)?;

            entry.set_offset(current_offset);
            let pad = pad::<8, u64>(entry.compressed_size());
            if pad > 0 {
                write_padding(&mut temp_file, pad as usize)?;
            }
            current_offset += entry.compressed_size() + pad;
        }

        // Write the index and footer
        let index_start = current_offset;
        for (name, entry) in &self.index {
            temp_file.write_all(entry.as_bytes())?;
            temp_file.write_all(name.as_bytes())?;
            let pad = pad::<BNDL_ALIGN, usize>(ENTRY_SIZE + name.len());
            if pad > 0 {
                write_padding(&mut temp_file, pad)?;
            }
        }

        let footer = Footer::new(index_start, self.index.len() as u32, FOOTER_MAGIC);
        temp_file.write_all(footer.as_bytes())?;
        temp_file.sync_all()?;

        // Acquire exclusive lock just before rename to prevent concurrent access
        self.file.lock_exclusive()?;

        // Release locks and close current file
        drop(self.mmap.take());
        let _ = self.file.unlock();

        // Atomically replace original with temp
        std::fs::rename(&temp_path, &self.path)?;

        // Reuse temp_file handle (still valid after rename)
        temp_file.lock_shared()?;
        let mmap = unsafe { Mmap::map(&temp_file)? };

        let footer_pos = mmap.len() - FOOTER_SIZE;
        let footer = Footer::read_from_bytes(&mmap[footer_pos..]).map_err(|_| {
            io::Error::new(
                io::ErrorKind::InvalidData,
                "Failed to read footer after vacuum",
            )
        })?;

        self.file = temp_file;
        self.mmap = Some(mmap);
        self.data_end = footer.index_offset();

        Ok(())
    }

    /// Reads an entry from the archive, decompressing if needed.
    ///
    /// Returns `None` if the entry doesn't exist or if CRC32 verification fails.
    pub fn read<'a>(&'a self, name: &str) -> Option<Cow<'a, [u8]>> {
        let entry = self.index.get(name)?;
        let mmap = self.mmap.as_ref()?;

        let data = if entry.compression_type() == Compress::Zstd {
            let compressed_data = mmap.get(
                entry.offset() as usize..(entry.offset() + entry.compressed_size()) as usize,
            )?;
            let mut out = Vec::with_capacity(entry.uncompressed_size() as usize);
            zstd::Decoder::new(compressed_data)
                .ok()?
                .read_to_end(&mut out)
                .ok()?;
            Cow::Owned(out)
        } else {
            let uncompressed_data = mmap.get(
                entry.offset() as usize..(entry.offset() + entry.uncompressed_size()) as usize,
            )?;
            Cow::Borrowed(uncompressed_data)
        };

        // Verify CRC32
        let computed_crc = crc32fast::hash(&data);
        if computed_crc != entry.crc32() {
            return None;
        }

        Some(data)
    }

    /// Reads an entry and writes it to the given writer.
    ///
    /// Returns the number of bytes written. Verifies CRC32 after reading.
    pub fn read_to<W: std::io::Write>(&self, name: &str, mut w: W) -> std::io::Result<u64> {
        let mut reader = self.reader(name)?;
        let bytes_copied = std::io::copy(&mut reader, &mut w)?;
        reader.verify_crc32()?;
        Ok(bytes_copied)
    }

    /// Returns a streaming reader for an entry.
    ///
    /// Automatically decompresses if the entry is compressed. Call [`Reader::verify_crc32()`] after reading to verify integrity.
    pub fn reader<'a>(&'a self, name: &str) -> io::Result<Reader<'a>> {
        let entry = self
            .index
            .get(name)
            .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "Entry not found"))?;

        let start = entry.offset() as usize;
        let end = start + entry.compressed_size() as usize;
        let mmap = self
            .mmap
            .as_ref()
            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidData, "Missing mmap"))?;
        let data_slice = &mmap[start..end];

        let cursor = io::Cursor::new(data_slice);

        if entry.compression_type() == Compress::Zstd {
            // Zstd streaming decoder
            let decoder = zstd::Decoder::new(cursor)?;
            Ok(Reader {
                decoder: Either::Left(decoder),
                crc32_hasher: Hasher::new(),
                expected_crc32: entry.crc32(),
            })
        } else {
            Ok(Reader {
                decoder: Either::Right(cursor),
                crc32_hasher: Hasher::new(),
                expected_crc32: entry.crc32(),
            })
        }
    }

    /// Returns the number of entries in the archive.
    pub fn len(&self) -> usize {
        self.index.len()
    }

    /// Returns true if the archive contains no entries.
    pub fn is_empty(&self) -> bool {
        self.index.is_empty()
    }

    /// Returns a reference to the archive index.
    ///
    /// The index maps entry names to their metadata.
    pub fn index(&self) -> &BTreeMap<String, Entry> {
        &self.index
    }

    /// Removes all entries from the index.
    ///
    /// Call [`save()`](Bindle::save) to commit. Data remains in the file until [`vacuum()`](Bindle::vacuum) is called.
    pub fn clear(&mut self) {
        self.index.clear()
    }

    /// Returns true if an entry with the given name exists.
    pub fn exists(&self, name: &str) -> bool {
        self.index.contains_key(name)
    }

    /// Removes an entry from the index.
    ///
    /// Returns true if the entry existed. Data remains in the file until [`vacuum()`](Bindle::vacuum) is called.
    pub fn remove(&mut self, name: &str) -> bool {
        self.index.remove(name).is_some()
    }

    /// Recursively adds all files from a directory to the archive.
    ///
    /// File paths are stored relative to the source directory. Call [`save()`](Bindle::save) to commit.
    pub fn pack<P: AsRef<Path>>(&mut self, src_dir: P, compress: Compress) -> io::Result<()> {
        self.pack_recursive(src_dir.as_ref(), src_dir.as_ref(), compress)
    }

    fn pack_recursive(
        &mut self,
        base: &Path,
        current: &Path,
        compress: Compress,
    ) -> io::Result<()> {
        if current.is_dir() {
            for entry in std::fs::read_dir(current)? {
                self.pack_recursive(base, &entry?.path(), compress)?;
            }
        } else {
            let name = current
                .strip_prefix(base)
                .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?
                .to_string_lossy();
            let mut data = Vec::new();
            File::open(current)?.read_to_end(&mut data)?;
            self.add(&name, &data, compress)?;
        }
        Ok(())
    }

    /// Extracts all entries to a destination directory.
    ///
    /// Creates subdirectories as needed to match the stored paths.
    pub fn unpack<P: AsRef<Path>>(&self, dest: P) -> io::Result<()> {
        let dest_path = dest.as_ref();
        if let Some(parent) = dest_path.parent() {
            std::fs::create_dir_all(parent)?;
        }
        for (name, _) in &self.index {
            if let Some(data) = self.read(name) {
                let file_path = dest_path.join(name);
                if let Some(parent) = file_path.parent() {
                    std::fs::create_dir_all(parent)?;
                }
                std::fs::write(file_path, data)?;
            }
        }
        Ok(())
    }

    /// Creates a streaming writer for adding an entry.
    ///
    /// The writer must be closed and then [`save()`](Bindle::save) must be called to commit the entry.
    pub fn writer<'a>(&'a mut self, name: &str, compress: Compress) -> io::Result<Writer<'a>> {
        self.file.lock_exclusive()?;
        self.file.seek(SeekFrom::Start(self.data_end))?;
        let compress = self.should_auto_compress(compress, 0);
        let f = self.file.try_clone()?;
        let start_offset = self.data_end;
        Ok(Writer {
            name: name.to_string(),
            bindle: self,
            encoder: if compress {
                Some(zstd::Encoder::new(f, 3)?)
            } else {
                None
            },
            start_offset,
            uncompressed_size: 0,
            crc32_hasher: Hasher::new(),
        })
    }
}

impl Drop for Bindle {
    fn drop(&mut self) {
        let _ = self.file.unlock();
    }
}
